src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -3046,9 +3227,14 @@ internal uint GetValEscape(BoundExpression expr, uint scopeOfTheContainingExpres
             // otherwise default to ExternalScope (ordinary values)
             switch (expr.Kind)
             {
+                case BoundKind.ThisReference:
+                    if (this.ContainingMember() is MethodSymbol { MethodKind: MethodKind.Constructor })
+                    {
+                        return Binder.ReturnOnlyScope;
+                    }
+                    return Binder.CallingMethodScope;
```

Comment:
Consider using `ThisParameter` similar to the corresponding code in `GetRefEscape()`:
```csharp
var thisParam = ((MethodSymbol)this.ContainingMember()).ThisParameter;
Debug.Assert(thisParam.Type.Equals(((BoundThisReference)expr).Type, TypeCompareKind.ConsiderEverything));
return GetParameterValEscape(thisParam);
```